cpp/dolfinx/fem/CoordinateElement.cpp

```diff
@@ -86,8 +95,110 @@ void CoordinateElement::compute_reference_geometry(
+    x0 = cell_geometry.matrix().transpose() * phi;
+
+    // Compute Jacobian and inverse
+    _finite_element->evaluate_reference_basis_derivatives(dphi_tensor, 1, X0);
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> J0(
+        _gdim, _tdim);
+    J0 = cell_geometry.matrix().transpose() * dphi;
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> K0(
+        _tdim, _gdim);
+    if (_gdim == _tdim)
+    {
+      K0 = J0.inverse();
+      detJ.fill(J0.determinant());
+    }
+    else
+    {
+      // Penrose-Moore pseudo-inverse
+      K0 = (J0.transpose() * J0).inverse() * J0.transpose();
+      detJ.fill(std::sqrt((J0.transpose() * J0).determinant()));
+    }
+
+    // Fill result for J, K and detJ
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Jview(J.data(), _gdim * num_points, _tdim);
+    Jview = J0.replicate(num_points, 1);
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Kview(K.data(), _tdim * num_points, _gdim);
+    Kview = K0.replicate(num_points, 1);
+
+    // Calculate X for each point
+    for (int ip = 0; ip < num_points; ++ip)
+      X.row(ip) = K0 * (x.row(ip).matrix().transpose() - x0);
+  }
+  else
+  {
+    // Newton's method for non-affine geometry
+    Eigen::VectorXd xk(x.cols());
+
+    for (int ip = 0; ip < num_points; ++ip)
+    {
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Jview(J.data() + ip * _gdim * _tdim, _gdim, _tdim);
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Kview(K.data() + ip * _gdim * _tdim, _tdim, _gdim);
+      Eigen::RowVectorXd Xk(_tdim);
+      // Xk = _reference_midpoint.head(_tdim);
+      // or Xk.setZero() ?
+      Xk.setZero();
+      const int max_its = 10;
+      int k;
+      for (k = 0; k < max_its; ++k)
+      {
+        // Compute physical coordinates
+        _finite_element->evaluate_reference_basis(phi_tensor, Xk);
+        xk = cell_geometry.matrix().transpose() * phi;
+
+        // Compute Jacobian and inverse
+        _finite_element->evaluate_reference_basis_derivatives(dphi_tensor, 1,
+                                                              Xk);
+        Jview = cell_geometry.matrix().transpose() * dphi;
+        if (_gdim == _tdim)
+          Kview = Jview.inverse();
+        else
+          Kview = (Jview.transpose() * Jview).inverse() * Jview.transpose();
+
+        // Increment to new point in reference
+        Eigen::VectorXd dX = Kview * (x.row(ip).matrix().transpose() - xk);
+        if (dX.squaredNorm() < 1e-12)
+          break;
+        Xk += dX;
+      }
+      if (k == max_its)
+      {
+        throw std::runtime_error(
+            "Iterations exceeded in Newton iteration for cmap");
```

Comment:
"cmap" is a bit cryptic here. Maybe something along "Newton method didn't converge for non-affine geometry"?

Reply:
Sure no problem to change